From 70a12837dcf3b8e02865e88f72db4f9195779e0b Mon Sep 17 00:00:00 2001
From: perillamint <perillamint@quendi.moe>
Date: Sun, 29 Apr 2018 15:48:06 +0900
Subject: [PATCH] Limit Nintendo Switch's external SD SDHCI bus clock speed.

This will fix Nintendo Switch does not load rootfs with some
"incompatible" SD cards, which supports UHS SDR104 speed.

So, this patch will solve this problem by limiting bus speed
to Switch's hardware can handle.

By experimentally obtained experience, 150MHz works fine with
64GB Samsung EVO+ microSD cards

However, 100MHz is more stable and supports more cards. Since Lakka doesn't
have a lot of I/O, 100Mhz is preferred over 150MHz.
---
 arch/arm64/boot/dts/nvidia/tegra210-nintendo-switch.dts | 1 +
 include/linux/mmc/card.h                                | 4 +++-
 2 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/include/linux/mmc/card.h b/include/linux/mmc/card.h
index 279b39008a33..b15af30d394c 100644
--- a/include/linux/mmc/card.h
+++ b/include/linux/mmc/card.h
@@ -151,7 +151,9 @@ struct sd_switch_caps {
 	unsigned int		hs_max_dtr;
 	unsigned int		uhs_max_dtr;
 #define HIGH_SPEED_MAX_DTR	50000000
-#define UHS_SDR104_MAX_DTR	208000000
+//#define UHS_SDR104_MAX_DTR	208000000
+// Blame faulty switch's SDHCI bus
+#define UHS_SDR104_MAX_DTR	100000000
 #define UHS_SDR50_MAX_DTR	100000000
 #define UHS_DDR50_MAX_DTR	50000000
 #define UHS_SDR25_MAX_DTR	UHS_DDR50_MAX_DTR
